Add `MULTISAMPLE_ARRAY` feature flag.

LaylBongers opened the pull request:
**Connections**
Closes #8565

**Description**
~~This adds a new `MULTISAMPLE_ARRAY` entry to `TextureFormatFeatureFlags`. When `TEXTURE_ADAPTER_SPECIFIC_FORMAT_FEATURES` is enabled and the feature is supported, applications can use this to create 2D Array textures and views with multi-sampling enabled.~~

This adds a new `MULTISAMPLE_ARRAY` entry to `Features`. When enabled and the feature is supported, applications can use this to create 2D Array textures and views with multi-sampling enabled

Additionally, this is implemented in the Vulkan backend. Following the Vulkan specificiation, this should be allowed:
https://docs.vulkan.org/refpages/latest/refpages/source/VkImageCreateInfo.html
https://docs.vulkan.org/refpages/latest/refpages/source/VkImageViewCreateInfo.html#VUID-VkImageViewCreateInfo-image-04972

*However* this is not necessarily supported if `VK_KHR_portability_subset` is used, which is the case for apple platform. In this case support should be detected from `multisampleArrayImage`.
https://docs.vulkan.org/refpages/latest/refpages/source/VkPhysicalDevicePortabilitySubsetFeaturesKHR.html
~~This isn't currently done yet, instead apple target, which always uses `VK_KHR_portability_subset` is for now excluded.~~
Now implemented too!

**To-Dos**
- [x] Feedback from someone more experienced with Vulkan. Am I missing something obvious here? Did I interpret the spec right?
- [x] Explicit checking for support in `VK_KHR_portability_subset` is not yet implemented.
- [x] ~~Currently `MULTISAMPLE_ARRAY` is marked true on all formats. In theory all formats that support multisample should also support this. Should this be more conservative?~~ Limited to the same degree as `MULTISAMPLE_RESOLVE`.
- [x] ~~Given that this bypasses some validation, are there any previous assumptions this breaks? Any things that now need to be explicitly checked?~~ Other limitations should already be caught by other constraints.
- [x] If there's low hanging fruit where it could be supported immediately on other platforms too without a lot of changes, that would be a nice-to-have.

**Testing**
I am actively testing this branch at https://github.com/celphase/indite. Developers with an OpenXR compatible headset should be able to test multiview there.

**Squash or Rebase?**
Squash

<!--
Thanks for filing! The codeowners file will automatically request reviews from the appropriate teams.

After you get a review and have addressed any comments, please explicitly re-request a review from the
person(s) who reviewed your changes. This will make sure it gets re-added to their review queue - you're not bothering us!
-->

**Checklist**

- [x] Run `cargo fmt`.
- [ ] Run `taplo format`.
- [x] Run `cargo clippy --tests`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
- [x] Run `cargo xtask test` to run tests.
- [x] If this contains user-facing changes, add a `CHANGELOG.md` entry. <!-- See instructions at the top of `CHANGELOG.md`. -->

LaylBongers committed:
Add `MULTISAMPLE_ARRAY` format feature flag.

LaylBongers committed:
Add apple feature gate to `MULTISAMPLE_ARRAY`, as it may not support it.

LaylBongers committed:
Limit `MULTISAMPLE_ARRAY` at least to the same degree as
`MULTISAMPLE_RESOLVE`.

LaylBongers committed:
Renumber flags to be more pretty ordered with `MULTISAMPLE_ARRAY`.

LaylBongers committed:
Add changes to CHANGELOG.md.

inner-daemons commented:
You asked about this in matrix, but you should probably still check the portability subset. With new drivers like KosmicKrisp and also HoneyKrisp, simply filtering for "apple" might not be a good idea. Also, other drivers can technically use these extensions.

LaylBongers commented:
While true that other drivers could use the portability subset, currently WGPU *only* requests a portability subset on apple, and does not store this in any way. Every other instance of code specifically for the portability extension uses this check. I could properly implement this check but that seems a bit out of scope if everything else does it this way.

LaylBongers committed:
Merge branch 'trunk' into trunk

teoxoy reviewed (changes requested):
I don't think this should be a format flag as none of the platforms make this conditional on the format. It should be an adapter/device feature.

Vulkan (non portability) and D3D12 support this unconditionally. Metal's [`MTLTextureType2DMultisampleArray`](https://developer.apple.com/documentation/metal/mtltexturetype/type2dmultisamplearray?language=objc) requires iOS/iPadOS 14.0+, macOS 10.14+, tvOS 16.0+, visionOS 1.0+. I haven't checked OpenGL ES.

Somewhat related: https://github.com/gfx-rs/wgpu/issues/4405 - though this is a different usecase.

LaylBongers committed:
Expose multisample array as an adapter feature instead of a format
feature.

LaylBongers committed:
Correct changelog for changes to PR.

LaylBongers committed:
Add `MULTISAMPLE_ARRAY` feature detection if `VK_KHR_portability_subset`
is active.

LaylBongers commented:
I've changed the feature to be an adapter flag instead.

This way it also was a lot easier to add support for portability platforms. I don't however have a device I can test that on.

LaylBongers committed:
Correct docs for `MULTISAMPLE_ARRAY` feature.

teoxoy reviewed (commented):
The implementation looks good but it should have at least a test making sure texture/view creation works and submitting a render pass with such an attachment works.

> If there's low hanging fruit where it could be supported immediately on other platforms too without a lot of changes, that would be a nice-to-have.

Exposing the feature unconditionally on D3D12 should be enough to get it going. Metal would need a bit more work (optional feature and changes to view creation).

LaylBongers committed:
Add unconditional support for `MULTISAMPLE_ARRAY` on DX12.

LaylBongers committed:
Add multiview multisample test.

LaylBongers committed:
Merge branch 'trunk' into trunk

LaylBongers commented:
I've added the DX12 platform, though I don't have a device that runs it so I can't test it.
I've added a multisample variant of the multiview tests, as they're the most likely use case for multisampled arrays anyways and it was the easiest to modify to add this.

LaylBongers commented:
I'm seeing the tests currently fail under DX12? I'll have to check the test in more detail tomorrow to see what exactly is failing.

LaylBongers commented:
I had a look at it just now and I'm a bit stumped on why the DX12 backend is failing on the test. I'm tempted to just turn off support in DX12 again for now.

inner-daemons commented on wgpu-core/src/device/resource.rs:
```diff
@@ -1436,7 +1436,9 @@ impl Device {
             }
 
-            if desc.size.depth_or_array_layers != 1 {
+            if desc.size.depth_or_array_layers != 1
+                && !self.features.contains(wgt::Features::MULTISAMPLE_ARRAY)
+            {
```
Looks very good

inner-daemons commented on wgpu-hal/src/vulkan/adapter.rs:
```diff
@@ -130,6 +130,11 @@ pub struct PhysicalDeviceFeatures {
+    /// Features provided by `VK_KHR_portability_subset`.
+    ///
+    /// Strictly speaking this tells us what features we *don't* have compared to core.
+    portability_subset: Option<vk::PhysicalDevicePortabilitySubsetFeaturesKHR<'static>>,
```
Does this need to be added to device creation?

LaylBongers replied:
Where specifically?

inner-daemons replied:
Check for references to other `*Features` members in this struct.

LaylBongers replied:
Ah good catch, added to the list.

inner-daemons commented on wgpu-types/src/features.rs:
```diff
@@ -1263,6 +1263,16 @@ bitflags_array! {
         const EXPERIMENTAL_MESH_SHADER_POINTS = 1 << 55;
+
+        /// Enables creating texture arrays that are also multisampled.
+        ///
+        /// Without this feature, you cannot create a texture that has both a `sample_count` higher
+        /// than 1, and a `depth_or_array_layers` higher than 1.
+        ///
+        /// Supported platforms:
+        /// - Vulkan (except VK_KHR_portability_subset if multisampleArrayImage is not available)
+        /// - DX12
+        const MULTISAMPLE_ARRAY = 1 << 56;
```
Any way we could add support to some Metal platforms in this PR? Especially with Apple Vision Pro this would probably be useful

LaylBongers replied:
It would be very useful for apple vision pro, but I don't have any apple products to test this on and the support seems to be non-trivial.

inner-daemons replied:
What about the support is non-trivial? Isn't it just checking if the OS version is high enough?

LaylBongers replied:
According to @teoxoy it would require a few more changes. I have absolutely zero experience with the Metal API.

inner-daemons replied:
@LaylBongers Alright. If you'd rather not than can you file an issue for Metal multisample array textures?

LaylBongers replied:
Added at #8593.

inner-daemons reviewed (changes requested):
Overall looks very good. If we could add support for Metal that'd be superb but you aren't obligated to do that.

I'll check out the CI failures & test later

LaylBongers committed:
Add missing `push_next` for `portability_subset`.

LaylBongers committed:
Merge branch 'trunk' into trunk

teoxoy commented:
I'm not sure why Windows is failing, I can't find anything obviously wrong but haven't debugged it too long either.

Feel free to remove advertising the feature on D3D12 to land this PR.

LaylBongers committed:
Remove multisample array from DX12 backend.

LaylBongers committed:
Merge branch 'trunk' into trunk

LaylBongers commented:
Looks like with DX12 removed, everything's building and testing fine :+1:

teoxoy commented on wgpu-types/src/features.rs:
```diff
@@ -1238,6 +1238,16 @@ bitflags_array! {
+        ///
+        /// Supported platforms:
+        /// - Vulkan (except VK_KHR_portability_subset if multisampleArrayImage is not available)
+        /// - DX12
```
```suggestion
```

LaylBongers replied:
Ah, my bad, completely forgot it was still listed!

LaylBongers committed:
Remove DX12 from list of supported platforms for `MULTISAMPLE_ARRAY`.

Co-authored-by: Teodor Tanasoaia <28601907+teoxoy@users.noreply.github.com>

inner-daemons commented:
Can you file an issue for support on DX12 @LaylBongers?
